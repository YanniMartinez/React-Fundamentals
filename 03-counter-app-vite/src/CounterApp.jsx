import PropTypes from 'prop-types';

export const CounterApp = ( { value } ) => {


  function hanbleAdd(event, newValue){ 
    // console.log(event);
    console.log(newValue);
  }

  return (
    <>
      <h1>CounterApp</h1>
      <h2>{ value }</h2>

      {/* Los argumentos pasan por defecto al handleAdd con propiedades
      de los eventos */}
      <button onClick={ (event) => handleAdd( event, 'Hola' )}>
        +1
      </button>
    </>
  )
}

CounterApp.propTypes={
    value: PropTypes.number.isRequired,
}

