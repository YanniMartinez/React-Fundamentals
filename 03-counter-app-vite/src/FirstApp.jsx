//import {Fragment} from 'react'
/**
 * Si no utilizamos la siguiente biblioteca : import {Fragment} from 'react'  entonces el código
 * se veria de la siguiente manera, generando un <div></div> de más
 * sumando procesamiento
 * 
    <div>
        <h1>Yanni Martinez</h1>
        <p>Soy un subtitulo</p>
    </div>
 * 
 * Sí usamos el fragmento entonces no debemos utilizarlo
 * 
 */
// export const FirstApp = () => {
//   return (
//     <Fragment>
//         <h1>Yanni Martinez</h1>
//         <p>Soy un subtitulo</p>
//     </Fragment>
//   )
// }

//Equivalente de fragmentos: Siempre debemos regresar almenos 1 nodo padre
//Podemos mostrar expresiones validas de JS pero no objetos
/**
 * Si no estamos modificando el valor y además no tiene relación con un hoock
 * entonces podemos ponerlo fuera del functional component, sin embargo, no se pon
 * en el scope global, significa que está dentro del componente encapsulado. 
 * REACT no re-renderizar elementos fuera de los functional components, ahorra recursos
 */
// const newMessage = {
//     message: 'Hola Mundo',
//     title: 'Yanni'
// };
// const newArray = [1,2,3,4,5,6,7,8,9];

/* const getResult = (a,b) =>{
    return a+b;
} */

//Importando las propTypes
import PropTypes from 'prop-types';


//Normalmente se desescturcturan los datos, como acontinuación:
export const FirstApp = ( {title, subTitle, name} ) => {

    return (
      <>
          {/* El data-testid sirve como un atributo para obtener valores */}
          <div data-testid="test-title">{ title }</div >
          <p>{ subTitle}</p>
          <p>{ name }</p>
      </>
    )
}


/**
 * Utilizando y definiendo las propTypes (Definimos un objeto)
 */
FirstApp.propTypes = {
    /* Le inficamos que el titulo debe ser string y adem´sa requerido */
    title: PropTypes.string.isRequired,
    subTitle: PropTypes.string.isRequired,
}


/**
 * Definiendo los default props
 */
FirstApp.defaultProps = {
    title: 'No hay titulo',
    subTitle: 'No hay Subtitulo',
    name: 'Yanni'
}

/*
export const FirstApp = ( props ) => {

    console.log(props);

    return (
      <>
          <h1>{ props.title }</h1>
          <p>Soy un subtitulo</p>
      </>
    )
}
 */