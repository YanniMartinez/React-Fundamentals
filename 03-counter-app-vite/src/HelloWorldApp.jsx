/**Declaración del functional component
 * Se les conoce así porque están basados en funciones
 * Las clases ya no son compatibles en REACT.
 * Siempre se tiene un componente
 * 
 * Si queremos exportar y que lo usen otros elementos debemos exportar primero
*/
export function App(){
    //document.createElement...
    return (<h1>Hola Mundo</h1>);
}


/**
 * Podria exportarse también de la siguiente manera:
 * function App(){
    return (<h1>Hola Mundo</h1>);
 }

    export default 
 * 
 */