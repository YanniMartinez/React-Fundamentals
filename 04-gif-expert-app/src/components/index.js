/** Esto es un archivo de BARRIL
 * Sirve para simplificar las lineas de importación
 * que tenemos dentro de nuestros códigos
 */
export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';
