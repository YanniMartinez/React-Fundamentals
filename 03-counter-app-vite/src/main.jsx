//Creando código HTML más XML
import React from 'react' //Importación de react
import ReactDOM from 'react-dom/client' //Herramienta para renderizar
import { FirstApp } from './FirstApp';
import { HelloWorldApp} from './HelloWorldApp'; 
import './styles.css'; //Importando los estilos globales

//Forma de realizar el Renderizando de la pagina
ReactDOM.createRoot(document.getElementById('root')).render(
    <React.StrictMode>
        {/* Comentario: 
        Mandando a llamar el nombre de nuestra function */}
        {/* <HelloWorldApp />  */}
        <FirstApp />
    </React.StrictMode>
);