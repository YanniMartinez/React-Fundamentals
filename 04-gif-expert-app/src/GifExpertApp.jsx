import { useState } from "react";
import { AddCategory, GifGrid } from "./components";

export const GifExpertApp = () =>{

    //Espacio en memoria para visualizar estados, en este caso le mandamos un arreglo de categorias
    const [categories, setcategories] = useState(['Dragon Ball','Smash bros']);

    /**
     * Agregar un nuevo elemento al estado de categorias
     */
    const onAddCategory = ( newCategory) =>{
        //categories.push('Zelda');
        if( categories.includes(newCategory) ) return;
        // setcategories( [...categories,'Zelda']);  //Forma 1 de hacerlo
        setcategories( [newCategory,...categories]);  //Forma 1 de hacerlo
        // setcategories( cat => [...cat, 'Valorant']); //Forma 2 de hacerlo
    }

    return(
        <>
            {/* Titulo */}
            <h1>GifExpertApp</h1>


            {/* Input */}
            {/* Le mandamos la referencia de las categorias */}
            <AddCategory 
                //setCategories={setcategories}
                onNewCategory={ value => onAddCategory(value) }
            />


            
            {/* Es forzoso mandar una llave */}
            { 
                categories.map( category =>
                (
                    <GifGrid key={category} 
                    category={category}/>
                )) 
            }

                {/* Gif Item */}
        </>
    );
}