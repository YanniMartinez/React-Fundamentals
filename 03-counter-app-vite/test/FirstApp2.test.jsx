/** @jest-environment jsdom */
import {render,screen} from '@testing-library/react'
import {FirstApp} from '../src/FirstApp'

/**El screen es el objeto que estamos renderizando */

describe('Pruebas en <FirstApp/>', () => {

  const title='Hola, Soy Goku';

  test('Debe de hacer match con el Snapshot', () => { 
    const{container}=render(<FirstApp title={title}/>);
    expect(container).toMatchSnapshot();
  })

  test('Debe de mostrar el mensaje "Hola, Soy Goku"',()=>{
    render(<FirstApp title={title}/>);
    screen.debug(); //Obtiene el estado del elemento, sirve para ver el objeto

    expect(screen.getByText(title)).toBeTruthy();

  })
  
})
