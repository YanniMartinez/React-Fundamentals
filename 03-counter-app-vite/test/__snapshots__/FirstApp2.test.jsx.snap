// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <FirstApp/> Debe de hacer match con el Snapshot 1`] = `
<div>
  <div
    data-testid="test-title"
  >
    Hola, Soy Goku
  </div>
  <p>
    No hay Subtitulo
  </p>
  <p>
    Yanni
  </p>
</div>
`;
